pkg/kubelet/pod_workers.go

```diff
@@ -31,14 +31,71 @@ import (
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+// OnCompleteFunc is a function that is invoked when an operation completes.
+// If err is non-nil, the operation did not complete successfully.
+type OnCompleteFunc func(err error)
+
+// PodStatusFunc is a function that is invoked to generate a pod status.
+type PodStatusFunc func(pod *api.Pod, podStatus *kubecontainer.PodStatus) api.PodStatus
+
+// UpdatePodOptions is an options struct to pass to a UpdatePod operation.
+type UpdatePodOptions struct {
+	// pod to update
+	Pod *api.Pod
+	// the mirror pod for the pod to update, if it is a static pod
+	MirrorPod *api.Pod
+	// the type of update (create, update, sync)
+	UpdateType kubetypes.SyncPodType
+	// optional callback function when operation completes
+	OnCompleteFunc OnCompleteFunc
+	// optional function to invoke to override pod status generation
+	PodStatusFunc PodStatusFunc
+	// optional override to use if a pod is being killed as part of update operation
+	// this should only ever be provided by hard eviction code paths otherwise pod data loss could occur
+	PodTerminationGracePeriodSecondsOverride *int64
+}
+
 // PodWorkers is an abstract interface for testability.
 type PodWorkers interface {
-	UpdatePod(pod *api.Pod, mirrorPod *api.Pod, updateType kubetypes.SyncPodType, updateComplete func())
+	UpdatePod(options *UpdatePodOptions)
 	ForgetNonExistingPodWorkers(desiredPods map[types.UID]empty)
 	ForgetWorker(uid types.UID)
 }
 
-type syncPodFnType func(*api.Pod, *api.Pod, *kubecontainer.PodStatus, kubetypes.SyncPodType) error
+// syncPodOptions provides the arguments to a SyncPod operation.
+type syncPodOptions struct {
+	// the mirror pod for the pod to sync, if it is a static pod
+	mirrorPod *api.Pod
+	// pod to sync
+	pod *api.Pod
+	// the current status
+	podStatus *kubecontainer.PodStatus
+	// optional function to invoke to override pod status generation
+	podStatusFunc PodStatusFunc
+	// optional override to use if a pod is being killed as part of update operation
+	podTerminationGracePeriodSecondsOverride *int64
+	// the type of update (create, update, sync)
+	updateType kubetypes.SyncPodType
+}
+
+// the function to invoke to perform a sync.
+type syncPodFnType func(options syncPodOptions) error
+
+// workUpdate is a value place-holder for the work queue
+type workUpdate struct {
```

Comment:
I had thought about consolidating `workUpdate` with `UpdatePodOptions`, I will look again to see why I did not go that route.  I think it was primarily how to do something that will be easier to digest in a review.  maybe that backfired in this case.